packages/wrangler/src/experimental-assets.ts

```diff
@@ -1,8 +1,301 @@
+import assert from "node:assert";
 import { existsSync } from "node:fs";
-import path from "node:path";
-import { UserError } from "./errors";
+import { readdir, readFile, stat } from "node:fs/promises";
+import * as path from "node:path";
+import chalk from "chalk";
+import { getType } from "mime";
+import PQueue from "p-queue";
+import prettyBytes from "pretty-bytes";
+import { fetchResult } from "./cfetch";
+import { formatTime } from "./deploy/deploy";
+import { FatalError, UserError } from "./errors";
+import { logger, LOGGER_LEVELS } from "./logger";
+import { hashFile } from "./pages/hash";
+import { isJwtExpired } from "./pages/upload";
+import { APIError } from "./parse";
+import { urlSafe } from "./sites";
 import type { Config } from "./config";
 
+export type AssetManifest = { [path: string]: { hash: string; size: number } };
+
+type InitializeAssetsResponse = {
+	// string of file hashes per bucket
+	buckets: string[][];
+	jwt: string;
+};
+
+export type UploadPayloadFile = {
+	name: string;
+	hash: string;
+	contents: string;
+	metadata: {
+		contentType: string;
+	};
+};
+
+type UploadResponse = {
+	jwt?: string;
+};
+
+// constants same as Pages for now
+const BULK_UPLOAD_CONCURRENCY = 3;
+const MAX_ASSET_COUNT = 20_000;
+const MAX_ASSET_SIZE = 25 * 1024 * 1024;
+const MAX_UPLOAD_ATTEMPTS = 5;
+const MAX_UPLOAD_GATEWAY_ERRORS = 5;
+
+export const syncExperimentalAssets = async (
+	accountId: string | undefined,
+	scriptName: string,
+	assetDirectory: string | undefined,
+	dryRun: boolean | undefined
+): Promise<string | undefined> => {
+	if (assetDirectory === undefined) {
+		return;
+	}
+	if (dryRun) {
+		logger.log("(Note: doing a dry run, not uploading or deleting anything.)");
+		return;
+	}
+	assert(accountId, "Missing accountId");
+
+	// 1. generate asset manifest
+	logger.info("🌀 Building list of assets...");
+	const manifest = await walk(assetDirectory, {});
+
+	// 2. fetch buckets w/ hashes
+	logger.info("🌀 Uploading asset manifest...");
```

Comment:
Users don't need to know about manifests :)

```suggestion
	logger.info("🌀 Starting asset upload...");
```